File: src/Java/testapp/src/main/java/com/myservicebus/testapp/dashboard/DashboardMetricsFilters.java
package com.myservicebus.testapp.dashboard;

import com.myservicebus.ConsumeContext;
import com.myservicebus.Filter;
import com.myservicebus.MessageUrn;
import com.myservicebus.Pipe;
import com.myservicebus.SendContext;

import java.net.URI;
import java.util.concurrent.CompletableFuture;

public final class DashboardMetricsFilters {
    private DashboardMetricsFilters() {
    }

    public static final class PublishMetricsFilter implements Filter<SendContext> {
        private final DashboardState state;

        public PublishMetricsFilter(DashboardState state) {
            this.state = state;
        }

        @Override
        public CompletableFuture<Void> send(SendContext context, Pipe<SendContext> next) {
            URI destination = context.getDestinationAddress();
            state.recordPublished(
                    destination == null ? null : DashboardSnapshotFactory.resolveMessageTypeFromDestination(destination),
                    destination == null ? null : DashboardSnapshotFactory.resolveMessageUrnFromDestination(destination));
            return next.send(context);
        }
    }

    public static final class SendMetricsFilter implements Filter<SendContext> {
        private final DashboardState state;

        public SendMetricsFilter(DashboardState state) {
            this.state = state;
        }

        @Override
        public CompletableFuture<Void> send(SendContext context, Pipe<SendContext> next) {
            URI destination = context.getDestinationAddress();
            state.recordSent(
                    destination == null ? null : DashboardSnapshotFactory.resolveMessageTypeFromDestination(destination),
                    destination == null ? null : DashboardSnapshotFactory.resolveMessageUrnFromDestination(destination));
            return next.send(context);
        }
    }

    public static final class ConsumeMetricsFilter<T> implements Filter<ConsumeContext<T>> {
        private final DashboardState state;
        private final String queueName;
        private final Class<T> messageClass;

        public ConsumeMetricsFilter(DashboardState state, String queueName, Class<T> messageClass) {
            this.state = state;
            this.queueName = queueName;
            this.messageClass = messageClass;
        }

        @Override
        public CompletableFuture<Void> send(ConsumeContext<T> context, Pipe<ConsumeContext<T>> next) {
            DashboardState.ConsumeScope scope = state.trackConsume(queueName, messageClass.getName(), MessageUrn.forClass(messageClass));
            return next.send(context)
                    .thenRun(scope::markSuccess)
                    .whenComplete((ignored, throwable) -> {
                        if (throwable != null) {
                            scope.markFault();
                        }
                        scope.close();
                    });
        }
    }
}

File: src/TestApp/DashboardMetricsFilters.cs
using MyServiceBus;

namespace TestApp;

public sealed class DashboardPublishMetricsFilter : IFilter<PublishContext>
{
    private readonly DashboardState _state;

    public DashboardPublishMetricsFilter(DashboardState state)
    {
        _state = state;
    }

    public Task Send(PublishContext context, IPipe<PublishContext> next)
    {
        _state.RecordPublished(
            context.DestinationAddress is null ? null : DashboardSnapshotFactory.ResolveMessageTypeFromDestination(context.DestinationAddress),
            context.DestinationAddress is null ? null : DashboardSnapshotFactory.ResolveMessageUrnFromDestination(context.DestinationAddress));
        return next.Send(context);
    }
}

public sealed class DashboardSendMetricsFilter : IFilter<SendContext>
{
    private readonly DashboardState _state;

    public DashboardSendMetricsFilter(DashboardState state)
    {
        _state = state;
    }

    public Task Send(SendContext context, IPipe<SendContext> next)
    {
        _state.RecordSent(
            context.DestinationAddress is null ? null : DashboardSnapshotFactory.ResolveMessageTypeFromDestination(context.DestinationAddress),
            context.DestinationAddress is null ? null : DashboardSnapshotFactory.ResolveMessageUrnFromDestination(context.DestinationAddress));
        return next.Send(context);
    }
}

public sealed class DashboardConsumeMetricsFilter<TMessage> : IFilter<ConsumeContext<TMessage>>
    where TMessage : class
{
    private readonly DashboardState _state;
    private readonly string _queueName;

    public DashboardConsumeMetricsFilter(DashboardState state, string queueName)
    {
        _state = state;
        _queueName = queueName;
    }

    public async Task Send(ConsumeContext<TMessage> context, IPipe<ConsumeContext<TMessage>> next)
    {
        using var scope = _state.TrackConsume(_queueName, typeof(TMessage).FullName ?? typeof(TMessage).Name, MessageUrn.For(typeof(TMessage)));
        try
        {
            await next.Send(context);
            scope.MarkSuccess();
        }
        catch
        {
            scope.MarkFault();
            throw;
        }
    }
}
